src/cluster/agglomerative.rs

```diff
@@ -0,0 +1,392 @@
+//!            &[4.6, 3.4, 1.4, 0.3],
+//!            &[5.0, 3.4, 1.5, 0.2],
+//!            &[4.4, 2.9, 1.4, 0.2],
+//!            &[4.9, 3.1, 1.5, 0.1],
+//!            &[7.0, 3.2, 4.7, 1.4],
+//!            &[6.4, 3.2, 4.5, 1.5],
+//!            &[6.9, 3.1, 4.9, 1.5],
+//!            &[5.5, 2.3, 4.0, 1.3],
+//!            &[6.5, 2.8, 4.6, 1.5],
+//!            &[5.7, 2.8, 4.5, 1.3],
+//!            &[6.3, 3.3, 4.7, 1.6],
+//!            &[4.9, 2.4, 3.3, 1.0],
+//!            &[6.6, 2.9, 4.6, 1.3],
+//!            &[5.2, 2.7, 3.9, 1.4],
+//!            ]);
+//!
+//! // Fit to data, with a threshold
+//! // example using FastPair, compute dendrogram
+//! let clustering = AggregativeFastPair::fit(&x).unwrap();
+//! let dendrogram = clustering.labels();
+//! ```
+//!
+use std::marker::PhantomData;
+
+use crate::algorithm::neighbour::fastpair::FastPair;
+use crate::error::Failed;
+use crate::linalg::Matrix;
+use crate::math::distance::euclidian::Euclidian;
+use crate::math::num::RealNumber;
+
+///
+/// Abstract trait for sequential, agglomerative, hierarchic, non-overlapping methods
+///
+pub trait SAHNClustering<T: RealNumber, M: Matrix<T>> {
+    ///
+    /// Aggregate the data according to given distance threshold
+    ///
+    fn fit(data: &M) -> Result<AggregativeFastPair<T, M>, Failed>;
+
+    ///
+    /// Return clusters labels
+    ///
+    fn labels(&self) -> &Vec<Vec<T>>;
+}
+
+///
+/// An implementation of Bottom-Up (Agglomerative) Hierarchical
+///  Clustering with `FastPair`
+///
+pub struct AggregativeFastPair<T: RealNumber, M: Matrix<T>> {
+    dendrogram: Vec<Vec<T>>, // computed labels
+    _marker: PhantomData<M>,
+}
+
+// Add linkage algorithms
+impl<T: RealNumber, M: Matrix<T>> FastCluster<T> for AggregativeFastPair<T, M> {}
+
+// Implement aggregative clustering using FastPair and fastcluster
+impl<T: RealNumber, M: Matrix<T>> SAHNClustering<T, M> for AggregativeFastPair<T, M> {
+    //
+    // 1. Compute `FastPair` on matrix's rows
+    // 2. Compute sparse dissimilarities into upper-trinagular matrix
+    // 3. Compute dendrogram
+    //
+    // The linkage distance threshold above which clusters will not be merged.
+    fn fit(data: &M) -> Result<AggregativeFastPair<T, M>, Failed> {
+        let fastpair = FastPair(data).unwrap();
+
+        // compute full connectivity from sparse matrix
+        let full_connectivity: M =
+            AggregativeFastPair::<T, M>::condensed_matrix(*(fastpair.connectivity.unwrap()), data);
+
+        // compute clusters
+        let dendrogram: Vec<Vec<T>> =
+            AggregativeFastPair::<T, M>::mst_single_linkage(full_connectivity, data.shape().0);
+
+        let n: usize = dendrogram.len();
+
+        Ok(AggregativeFastPair {
+            // assign computed dendrogram to structure
+            dendrogram: AggregativeFastPair::<T, M>::label(dendrogram, n),
+            _marker: PhantomData,
+        })
+    }
+
+    fn labels(&self) -> &Vec<Vec<T>> {
```

Comment:
we need 2 methods that calculate cluster labels. Once will calculate labels for fixed number of clusters, k, while the other method produce labels from a threshold parameter. Take a look at the Jupyter notebook with demonstration that I've attached to this PR. 

I am not sure what would be the best way to call these methods, `labels_by_k`, `labels_by_threshold`?

Reply:
I can do it but I need the pseudo-code for the methods mentioned, as I have no idea how to compute the outcome from the dendrogram. The current implementation compute the dense dendrogram in the shape (for the Iris dataset):
```
       let expected = [
            //  index, index, distance, label
            [0.0, 4.0, 0.01999999999999995, 2.0],
            [7.0, 19.0, 0.029999999999999964, 3.0],
            [1.0, 9.0, 0.030000000000000037, 2.0],
            [14.0, 18.0, 0.05999999999999993, 2.0],
            [2.0, 3.0, 0.0600000000000001, 2.0],
            [6.0, 23.0, 0.06999999999999998, 3.0],
            [10.0, 12.0, 0.07000000000000003, 2.0],
            [11.0, 16.0, 0.07000000000000013, 2.0],
            [8.0, 24.0, 0.08999999999999982, 4.0],
            [21.0, 27.0, 0.09000000000000011, 6.0],
            [20.0, 28.0, 0.10999999999999982, 9.0],
            [22.0, 26.0, 0.1799999999999998, 4.0],
            [25.0, 30.0, 0.2600000000000009, 6.0],
            [13.0, 29.0, 0.27000000000000013, 10.0],
            [5.0, 32.0, 0.3800000000000002, 11.0],
            [15.0, 33.0, 0.54, 12.0],
            [31.0, 34.0, 0.6899999999999995, 18.0],
            [17.0, 35.0, 0.7000000000000001, 19.0],
            [5.0, 17.0, 4.460000000000001, 0.0],
        ];
```
From this the methods should compute the "by cluster number" and "by threshold", can you provide the procedures to do so or provide a link to code?

Reply:
@Mec-iS Take a look at how it is done in Scikit's [`_agglomerative.py`](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/cluster/_agglomerative.py). Start from [_hc_cut method](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/cluster/_agglomerative.py#L632). If I understand this method correctly, the basic idea is to take k records from the dendrogram with the smallest index values - these are your top nodes in k clusters.  All you need to do next is to iterate through descendants of each top node and assign all nodes at the lower level label of its parent. 
Graphically it will look as if you are making a horizontal cut at some level that covers exactly k trees. For example, on this picture 
![image](https://user-images.githubusercontent.com/1102497/101837635-35901b80-3af4-11eb-98d0-b21324288906.png)
the horizontal cut will produce k=3 clusters. Now you need to follow descendants of these three nodes and assign them labels of their parent. 

Once you are done with "by cluster number" you can easily implement method "by threshold" using [this logic](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/cluster/_agglomerative.py#L909)